snakemake/__init__.py

```diff
@@ -2286,6 +2290,20 @@ def get_argument_parser(profile=None):
         "that is compatible with (or ideally the same as) the currently "
         "running version.",
     )
+    group_kubernetes.add_argument(
+        "--k8s-cpu-scalar",
+        metavar="FLOAT",
+        default=1,
+        type=float,
+        help="K8s reserves some proportion of available CPUs for its own use. "
+        "So, where an underlying node may have 8 CPUs, only e.g. 7600 milliCPUs "
+        "are allocatable to k8s pods (i.e. snakemake jobs). As 8 > 7.6, k8s can't "
+        "find a node with enough CPU resource to run such jobs. This argument acts "
+        "as a global scalar on each job's CPU request, so that e.g. a job whose "
+        "rule definition asks for 8 CPUs will request 7600m CPUs from k8s, "
+        "allowing it to utilise one entire node. N.B: the job itself would still "
+        "see the original value, i.e. as the value substituted in {threads}.",
+    )
```

Comment:
Very good catch! Can this fraction be somehow queried from kubernetes API instead of requiring the user to set it? Or alternatively, can we set it to a reasonable default that will work at least on all k8s clusters?

Reply:
So I believe there is a way to query how much CPU resource is assignable per node, but that is a "live" metric (basically N - `${currently_assigned_mCPUs}`) which will change depending on how many pods are running on the node. During execution, those pods will be a mix of system pods and snakemake jobs, so we don't want to use that number. It would also vary a little between nodes, and we don't know which node this job will be scheduled on.

So, I think it is best to set a global default rather than query k8s. I don't think there is a single value across all K8S deployments, but at least in my experience 0.95 would be a reasonably safe default. The other option would be that we always subtract e.g. 500mCPUs (0.5 of a CPU) from all nodes, hoping that that is enough to catch all system pods, but that seems like a bad idea as we would waste (proportionally) more with smaller nodes and still not be guaranteed to solve the original problem.

Reply:
Ok, I trust your expertise here, and we go for the 0.95 as a default with this parameter.
```suggestion
    group_kubernetes.add_argument(
        "--k8s-cpu-scalar",
        metavar="FLOAT",
        default=0.95,
        type=float,
        help="K8s reserves some proportion of available CPUs for its own use. "
        "So, where an underlying node may have 8 CPUs, only e.g. 7600 milliCPUs "
        "are allocatable to k8s pods (i.e. snakemake jobs). As 8 > 7.6, k8s can't "
        "find a node with enough CPU resource to run such jobs. This argument acts "
        "as a global scalar on each job's CPU request, so that e.g. a job whose "
        "rule definition asks for 8 CPUs will request 7600m CPUs from k8s, "
        "allowing it to utilise one entire node. N.B: the job itself would still "
        "see the original value, i.e. as the value substituted in {threads}.",
    )
```